argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+	}
+}
+
+// NoArgon2OutOfProcessWatchdogMonitor is an implmenentation of Argon2OutOfProcessWatchdogMonitor that
+// provides no watchdog functionality. It is paired with [NoArgon2OutOfProcessWatchdogHandler] on the
+// remote side. It holds the watchdog goroutine in a parked state for the lifetime of the tomb.
+func NoArgon2OutOfProcessWatchdogMonitor() Argon2OutOfProcessWatchdogMonitor {
+	return func(tmb *tomb.Tomb, reqChan chan<- *Argon2OutOfProcessRequest, rspChan <-chan *Argon2OutOfProcessResponse) error {
+		select {
+		case <-tmb.Dying():
+			return tomb.ErrDying
+		case <-rspChan:
+			return errors.New("unexpected watchdog response")
+		}
+	}
+}
+
+type outOfProcessArgon2KDFImpl struct {
+	newHandlerCmd func() (*exec.Cmd, error)
+	timeout       time.Duration
+	watchdog      Argon2OutOfProcessWatchdogMonitor
+}
+
+func (k *outOfProcessArgon2KDFImpl) sendRequestAndWaitForResponse(req *Argon2OutOfProcessRequest) (*Argon2OutOfProcessResponse, error) {
+	// Use ther user-supplied function to create a new *exec.Cmd structure.
+	cmd, err := k.newHandlerCmd()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create new command: %w", err)
+	}
+
+	// Configure an OS pipe for stdin for sending requests.
+	stdinPipe, err := cmd.StdinPipe()
+	if err != nil {
+		// This doesn't fail once the OS pipe is created, so there's no
+		// cleanup to do on failure paths.
+		return nil, fmt.Errorf("cannot create stdin pipe: %w", err)
+	}
+
+	// Configure an OS pipe for stdout for receiving responses.
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		// This doesn't fail once the OS pipe is created, so there's no
+		// cleanup to do on failure paths other than closing the stdinPipe
+		stdinPipe.Close()
+		return nil, fmt.Errorf("cannot create stdout pipe: %w", err)
+	}
+
+	// Start the remote process.
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("cannot start handler process: %w", err)
+	}
+
+	var actualRsp *Argon2OutOfProcessResponse // The response to return to the caller
+	exitWaitCh := make(chan struct{})         // A channel which signals successful exit of the child process when closed
+	tmb := new(tomb.Tomb)                     // To track all goroutines
+
+	// Spin up a routine to bootstrap the parent side and handle responses from
+	// the remote process.
+	tmb.Go(func() error {
+		// Spin up a routine for killing the child process if it doesn't
+		// die cleanly
+		tmb.Go(func() error {
+			<-tmb.Dying() // Wait here until the tomb enters a dying state
+
+			select {
+			case <-exitWaitCh:
+				// The command closed cleanly, so there's nothing to do.
+				return tomb.ErrDying
+			case <-time.NewTimer(5 * time.Second).C:
+				// We've waited 5 seconds - kill the child process instead. Go 1.20
+				// has a new feature (WaitDelay) which might make things a bit better
+				// here because I don't know how racey things are here - exec.Cmd is
+				// quite complicated.
+				if err := cmd.Process.Kill(); err != nil {
```

Comment:
Depends on how much we trust the caller to have implemented things correctly, but if they implemented the runner their own binary it may be worth to at least leave a comment somewhere that the command should likely be run in a new process group and kill should be done through syscall.Kill(-pgid, syscall.SIGKILL). In which case, some part of the code should set `Setpgid: true` in `cmd.SysProcAttr`.